pkg/allocation/converters/converter.go

```diff
@@ -77,13 +91,12 @@ func ConvertGSAToAllocationRequest(in *allocationv1.GameServerAllocation) *pb.Al
 	}
 
 	out := &pb.AllocationRequest{
-		Namespace:                    in.GetNamespace(),
-		PreferredGameServerSelectors: convertInternalLabelSelectorsToLabelSelectors(in.Spec.Preferred),
-		Scheduling:                   convertGSASchedulingStrategyToAllocationScheduling(in.Spec.Scheduling),
+		Namespace:  in.GetNamespace(),
+		Scheduling: convertGSASchedulingStrategyToAllocationScheduling(in.Spec.Scheduling),
 		MultiClusterSetting: &pb.MultiClusterSetting{
 			Enabled: in.Spec.MultiClusterSetting.Enabled,
 		},
-		RequiredGameServerSelector: convertInternalGameServerSelectorToGameServer(&in.Spec.Required),
+		GameServerSelectors: convertInternalLabelSelectorsToLabelSelectors(in.Spec.Selectors),
```

Comment:
It looks like `ConvertGSAToAllocationRequest` is only used from `allocateFromRemoteCluster` so this won't change any user visible behavior to only use the new fields here.

However, it does mean that if you are using multi-cluster allocation and you have some clusters at an old version and some at a new version that the cross cluster allocation requests will start to fail. 

I'm wondering if we should actually keep using the old fields here for 1-2 releases, and then make this change. That would allow folks to do multi-cluster allocation with agones versions +/- at least 1 version without any trouble. Making this change effectively requires all clusters doing multi-cluster allocation to be upgraded to have the new CRD & allocation code simultaneously. 

@markmandel  WDYT?

Reply:
As long as users do not use the new filed selectors, the multi-cluster allocation should not fail, and yes, if the user would like to use selectors field for mismatched agones versions (<1.17 & >1.17), the old clusters won't accept selectors and will error out that the required field is not set.

Reply:
I'm just trying to make sure I wrap my head around this correctly, because it is a fun one.

| version ➡️  version | works? | 
|--|-|
| 1.17 ➡️ 1.17 | works! (b/c same everything) |
| 1.16 ➡️  1.17 | works! (b/c 1.16 will send deprecated fields, which 1.17 will convert) |
| 1.17 ➡️  1.16 | FAILS! (b/c 1.16 will receive the new fields, and not know what to do with them)

Did I get that right?

Reply:
@markmandel - that is the way I read it as well. 

My suggestion to keep using the old fields here (and actually - maybe converting selectors into the old fields) would allow 1.17 -> 1.16 to work. Then we can switch to this code in 1.18, because at that point we'd have 1.17 -> 1.18 using the old fields (which would work) and 1.18 -> 1.17 using the new fields (which would also work, since the new fields exist in 1.17). 

It depends on how much we care about supporting MCA in a compatible way between versions. We don't really advertise a compatibility matrix for MCA so maybe just making sure that 1.x -> 1.x works is sufficient?

Reply:
> My suggestion to keep using the old fields here (and actually - maybe converting selectors into the old fields) would allow 1.17 -> 1.16 to work. Then we can switch to this code in 1.18, because at that point we'd have 1.17 -> 1.18 using the old fields (which would work) and 1.18 -> 1.17 using the new fields (which would also work, since the new fields exist in 1.17).

Yeah, I'm not sure either. 🤔  assuming that we do keep the old fields - are we going to test the MCA works across versions? (have we ever tested this?). I'm leaning towards maintaining backward compatibility, but also happy to defer to another strong opinion.

(I think what makes me feel like it's icky, is that would we need to go from selector -> required and preferred, and for whatever reason that feels like it's wrong 😄)

@cindy52 does the new conversion pieces make this easier/harder?

Random architecture thought that I think makes this easier (I think this works, let me know):

What if `GameServerAllocation.Converter()` didn't remove the preferred and required values? and if a person supplied both, we didn't validate it (or at least for a version) - then we could easily everything to the next MCA - required/preferred and selectors - and it could decide what made sense for it. 

Maybe in 1.18 (or at some point) we could then add validation to say you can't have `selectors` and also `required` and/or `preferred`, and can stop sending all fields?

Another question: I assume we don't mind if we break things if someone is upgrading from 1.17 to 1.21 (way in the future)?

Reply:
I think no matter when we add the validation, it will break the upgrade at some point. Imagine a cluster is upgraded to the version that will execute the validation, the persistent gameserverallocation crds from old version (if contains both selectors and preferred/required fields) will fail. Not sure if it's encouraged to keep the crds from old version when upgrade in Agones.

Reply:
> I think no matter when we add the validation, it will break the upgrade at some point.

That is an excellent point. 🤔 hmnnn...

So we either

1. Don't ever have validation on required/preferred at the same time as selector
2. We break intra-version MCA for a release.

Sounds like the two options?

Reply:
Added  "hack-selectors-empty" matchlabel to the end of preferred to indicate that the selectors field was empty but converted by required&preferred so later when convert back, the selectors field won't be added.